File: cpp/src/steering.cpp
/**
 * @file steering.cpp
 * @brief Steering magnet class implementation
 * @author Hirokazu Maesaka
 * @date 2025
 */
// steering.cpp
//
// Copyright (C) 2025 Hirokazu Maesaka (RIKEN SPring-8 Center)
//
// This file is part of Egret: Engine for General Research in
// Energetic-beam Tracking.
//
// Egret is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include "egret/steering.hpp"
#include <ranges>

/**
 * @brief Compute effective kick angles considering tilt and momentum deviation.
 * @param delta Relative momentum deviation
 * @return std::tuple<double, double> Effective kick angles (kick_x_eff, kick_y_eff)
 */
std::tuple<double, double> egret::Steering::tilted_kick(const double delta) const noexcept {
    const double cos_tilt = std::cos(tilt_);
    const double sin_tilt = std::sin(tilt_);
    const double denom = 1.0 + delta;
    const double kick_x_eff = (kick_x_ * cos_tilt - kick_y_ * sin_tilt) / denom;
    const double kick_y_eff = (kick_x_ * sin_tilt + kick_y_ * cos_tilt) / denom;
    return std::make_tuple(kick_x_eff, kick_y_eff);
}

/**
 * @brief Calculate the additive dispersion function at the end of the steering magnet.
 * @param cood0 Optional input coordinate
 * @param ds Step size (unused)
 * @param method Integration method (unused)
 * @return Eigen::Vector4d Additive dispersion vector [eta_x, eta'_x, eta_y, eta'_y]
 */
Eigen::Vector4d egret::Steering::dispersion(const std::optional<Coordinate> &cood0,
    const double ds, IntegrationMethod method) const noexcept(false) {
    (void)ds; // unused parameter
    (void)method; // unused parameter
    const double delta = cood0 ? cood0->delta() : 0.0;
    const auto [kick_x_eff, kick_y_eff] = tilted_kick(delta);
    const double eta_x = -0.5 * length_ * kick_x_eff;
    const double eta_y = -0.5 * length_ * kick_y_eff;
    const double etap_x = -kick_x_eff;
    const double etap_y = -kick_y_eff;
    Eigen::Vector4d eta;
    eta << eta_x, etap_x, eta_y, etap_y;
    return eta;
}

/**
 * @brief Calculate the additive dispersion function array through the steering magnet.
 * @param cood0 Optional input coordinate
 * @param ds Step size
 * @param endpoint Whether to include the endpoint
 * @param method Integration method (unused)
 * @return std::tuple<Eigen::Matrix<double, 4, Eigen::Dynamic>, Eigen::ArrayXd> Array of additive dispersion vectors and corresponding s values
 */
std::tuple<Eigen::Matrix<double, 4, Eigen::Dynamic>, Eigen::ArrayXd>
egret::Steering::dispersion_array(const std::optional<Coordinate> &cood0,
    const double ds, const bool endpoint, IntegrationMethod method) const noexcept(false) {
    (void)method; // unused parameter
    const auto s_array = Element::s_array(ds, endpoint); // ArrayXd
    const size_t n = s_array.size();
    const double delta = cood0 ? cood0->delta() : 0.0;
    const auto [kick_x_eff, kick_y_eff] = tilted_kick(delta);
    Eigen::Matrix<double, 4, Eigen::Dynamic> eta(4, n);
    if (n == 1) {
        eta << 0.0, -kick_x_eff, 0.0, -kick_y_eff;
    } else if (std::abs(length_) > 0.) {
        const auto eta_x = -0.5 * kick_x_eff * s_array * s_array / length_; // ArrayXd
        const auto eta_y = -0.5 * kick_y_eff * s_array * s_array / length_; // ArrayXd
        const auto etap_x = -kick_x_eff * s_array / length_; // ArrayXd
        const auto etap_y = -kick_y_eff * s_array / length_; // ArrayXd
        eta.row(0) = eta_x;
        eta.row(1) = etap_x;
        eta.row(2) = eta_y;
        eta.row(3) = etap_y;
    } else {
        throw std::runtime_error("Steering length is zero and array length is more than 1 in dispersion_array calculation.");
    }
    return std::make_tuple(eta, s_array);
}

/**
 * @brief Transfer the coordinate, envelope, and dispersion through the steering magnet.
 * @param cood0 Initial coordinate
 * @param evlp0 Initial envelope (optional)
 * @param disp0 Initial dispersion (optional)
 * @param ds Step size for integration
 * @param method Integration method (unused)
 * @return std::tuple<egret::Coordinate, std::optional<egret::Envelope>, std::optional<egret::Dispersion>> Coordinate, envelope, and dispersion after transfer
 */
std::tuple<egret::Coordinate, std::optional<egret::Envelope>, std::optional<egret::Dispersion>>
egret::Steering::transfer(const Coordinate &cood0, const std::optional<Envelope> &evlp0,
    const std::optional<Dispersion> &disp0, const double ds, IntegrationMethod method) const noexcept(false) {
    (void)method; // unused parameter
    const double delta = cood0.delta();
    const auto [kick_x_eff, kick_y_eff] = tilted_kick(delta);
    const double dx = 0.5 * length_ * kick_x_eff;
    const double dy = 0.5 * length_ * kick_y_eff;
    const auto M = transfer_matrix(cood0, ds);
    const auto kick_vec = Eigen::Vector4d(dx, kick_x_eff, dy, kick_y_eff); // Vector4d
    const auto cood_vec = M * cood0.vector() + kick_vec;
    const Coordinate cood(cood_vec, cood0.s() + length_, cood0.z(), cood0.delta());
    std::optional<Envelope> evlp = evlp0;
    if (evlp) {
        evlp->transfer(M, length_);
    }
    std::optional<Dispersion> disp = disp0;
    if (disp) {
        const auto disp_add = dispersion(cood0, ds); // Vector4d
        const auto disp_vec = M * disp0->vector() + disp_add; // Vector4d
        disp->vector(disp_vec);
        disp->s(disp0->s() + length_);
    }
    return std::make_tuple(cood, evlp, disp);
}

/**
 * @brief Transfer the coordinate, envelope, and dispersion through the steering magnet for an array of steps.
 * @param cood0 Initial coordinate
 * @param evlp0 Initial envelope (optional)
 * @param disp0 Initial dispersion (optional)
 * @param ds Step size for integration
 * @param endpoint Whether to include the endpoint in the array
 * @param method Integration method (unused)
 * @return std::tuple<egret::CoordinateArray, std::optional<egret::EnvelopeArray>, std::optional<egret::DispersionArray>> Coordinate, envelope, and dispersion arrays after transfer
 */
std::tuple<egret::CoordinateArray, std::optional<egret::EnvelopeArray>, std::optional<egret::DispersionArray>>
egret::Steering::transfer_array(const Coordinate &cood0,
    const std::optional<Envelope> &evlp0,
    const std::optional<Dispersion> &disp0,
    const double ds, const bool endpoint,
    IntegrationMethod method) const noexcept(false) {
    (void)method; // unused parameter
    const double delta = cood0.delta();
    const auto [kick_x_eff, kick_y_eff] = tilted_kick(delta);
    const auto [M_array, s_array] = transfer_matrix_array(cood0, ds, endpoint); // vector<Matrix4d>, ArrayXd
    const size_t n = s_array.size();
    Eigen::Matrix<double, 4, Eigen::Dynamic> kick_vector_array(4, n);
    if (n == 1) {
        kick_vector_array << 0.0, kick_x_eff, 0.0, kick_y_eff;
    } else if (std::abs(length_) > 0.) {
        const auto dx_array = 0.5 * kick_x_eff * s_array; // ArrayXd
        const auto dy_array = 0.5 * kick_y_eff * s_array; // ArrayXd
        const auto kick_x_array = kick_x_eff * s_array / length_; // ArrayXd
        const auto kick_y_array = kick_y_eff * s_array / length_; // ArrayXd
        kick_vector_array.row(0) = dx_array;
        kick_vector_array.row(1) = kick_x_array;
        kick_vector_array.row(2) = dy_array;
        kick_vector_array.row(3) = kick_y_array;
    } else {
        throw std::runtime_error("Steering length is zero and array length is more than 1 in transfer_array calculation.");
    }
    Eigen::Matrix<double, Eigen::Dynamic, 4> M_combined(4 * n, 4);
    for (const size_t i : std::views::iota(0u, n)) {
        M_combined.block<4, 4>(4 * i, 0) = M_array[i];
    }
    const auto vector_array = (M_combined * cood0.vector()).reshaped(4, n)
        + kick_vector_array; // Matrix 4xn
    const CoordinateArray cood_array(vector_array, cood0.s() + s_array,
        Eigen::ArrayXd::Constant(n, cood0.z()),
        Eigen::ArrayXd::Constant(n, cood0.delta()));
    std::optional<EnvelopeArray> evlp_array;
    if (evlp0) {
        evlp_array = EnvelopeArray::transport(*evlp0, M_array, s_array);
    }
    std::optional<DispersionArray> disp_array;
    if (disp0) {
        const auto [disp_add_array, _] = dispersion_array(cood0, ds, endpoint); // Matrix 4xn, ArrayXd
        const auto disp_vector_array = (M_combined * disp0->vector()).reshaped(4, n)
            + disp_add_array; // Matrix 4xn
        disp_array = DispersionArray(disp_vector_array, disp0->s() + s_array);
    }
    return std::make_tuple(cood_array, evlp_array, disp_array);
}

File: egret/python/steering.py
# python/steering.py
#
# Copyright (C) 2025 Hirokazu Maesaka (RIKEN SPring-8 Center)
#
# This file is part of Egret: Engine for General Research in
# Energetic-beam Tracking.
#
# Egret is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations
from ..base.steering import Steering as SteeringABC
from .element import Element
from .coordinate import Coordinate
from .coordinatearray import CoordinateArray
from .envelope import Envelope
from .envelopearray import EnvelopeArray
from .dispersion import Dispersion
from .dispersionarray import DispersionArray
from .drift import Drift
import numpy as np
import numpy.typing as npt
from typing import Tuple

class Steering(SteeringABC, Element):
    '''
    Steering magnet class.
    '''

    def __init__(self, name: str, length: float,
                 kick_x: float = 0., kick_y: float = 0.,
                 dx: float = 0., dy: float = 0., ds: float = 0.,
                 tilt: float = 0., info: str = ''):
        '''
        Args:
            name str: Name of the element.
            length float: Length of the element [m].
            kick_x float: Horizontal kick angle [rad].
            kick_y float: Vertical kick angle [rad].
            dx float: Horizontal offset of the element [m].
            dy float: Vertical offset of the element [m].
            ds float: Longitudinal offset of the element [m].
            tilt float: Tilt angle of the element [rad].
            info str: Additional information.
        '''
        super().__init__(name, length, 0.0, dx, dy, ds, tilt, info)
        self._kick_x = kick_x
        self._kick_y = kick_y

    @property
    def kick_x(self) -> float:
        '''
        Horizontal kick angle [rad].
        '''
        return self._kick_x

    @property
    def kick_y(self) -> float:
        '''
        Vertical kick angle [rad].
        '''
        return self._kick_y

    @property
    def kick(self) -> Tuple[float, float]:
        '''
        Return a tuple of horizontal and vertical kick angles [rad].
        '''
        return self._kick_x, self._kick_y

    @kick_x.setter
    def kick_x(self, kick_x: float) -> None:
        '''
        Set horizontal kick angle.

        Args:
            kick_x float: Horizontal kick angle [rad].
        '''
        self._kick_x = kick_x

    @kick_y.setter
    def kick_y(self, kick_y: float) -> None:
        '''
        Set vertical kick angle.

        Args:
            kick_y float: Vertical kick angle [rad].
        '''
        self._kick_y = kick_y

    @kick.setter
    def kick(self, kick_x: float, kick_y: float) -> None:
        '''
        Set the steering angles.

        Args:
            kick_x float: Horizontal kick angle [rad].
            kick_y float: Vertical kick angle [rad].
        '''
        self._kick_x = kick_x
        self._kick_y = kick_y

    def copy(self) -> Steering:
        '''
        Return a copy of the steering magnet.

        Returns:
            Steering: Copied steering magnet.
        '''
        return Steering(self._name, self.length, self._kick_x, self._kick_y,
                        self._dx, self._dy, self._ds, self._tilt, self._info)

    def _tilted_kick(self, delta) -> Tuple[float, float]:
        '''
        Get the tilted kick angles.

        Args:
            delta float: Relative energy deviation.

        Returns:
            Tuple[float, float]: Tilted kick angles (dxp_tilted, dyp_tilted).
        '''
        cosphi, sinphi = np.cos(self.tilt), np.sin(self.tilt)
        rotmat = np.array([[cosphi, -sinphi], [sinphi, cosphi]])
        kick_tilted = np.dot(rotmat, np.array([self.kick_x, self.kick_y])) / (1. + delta)
        return kick_tilted[0], kick_tilted[1]

    def transfer_matrix(self, cood0: Coordinate = None, ds: float = 0.1, method: str = 'symplectic4') -> npt.NDArray[np.floating]:
        '''
        Transfer matrix of the steering magnet.

        Args:
            cood0 Coordinate: Initial coordinate (not used in the steering magnet).
            ds float: Maximum step size [m] for integration. (not used in the steering magnet).
            method str: Integration method. ('midpoint', 'rk4', 'symplectic{1,2,4}') (Not used in the steering magnet).

        Returns:
            npt.NDArray[np.floating]: 4x4 transfer matrix.
        '''
        return Drift.transfer_matrix_from_length(self._length)

    def transfer_matrix_array(self, cood0: Coordinate = None, ds: float = 0.1, endpoint: bool = False, method: str = 'symplectic4') \
        -> Tuple[npt.NDArray[np.floating], npt.NDArray[np.floating]]:
        '''
        Transfer matrix array along the steering magnet.

        Args:
            cood0 Coordinate: Initial coordinate (not used in the steering magnet).
            ds float: Maximum step size [m].
            endpoint bool: If True, include the endpoint.
            method str: Integration method. ('midpoint', 'rk4', 'symplectic{1,2,4}') (Not used in the steering magnet).

        Returns:
            npt.NDArray[np.floating]: Transfer matrix array of shape (4, 4, N).
            npt.NDArray[np.floating]: Longitudinal positions [m].
        '''
        return Drift.transfer_matrix_array_from_length(self._length, ds, endpoint)

    def dispersion(self, cood0: Coordinate, ds: float = 0.1, method: str = 'symplectic4') -> npt.NDArray[np.floating]:
        '''
        Additive dispersion function at the end of the steering magnet.

        Args:
            cood0 Coordinate: Initial coordinate. (Only delta is used here.)
            ds float: Maximum step size for integration [m]. (Not used in this method.)
            method str: Integration method. ('midpoint', 'rk4', 'symplectic{1,2,4}') (Not used in the Steering class.)

        Returns:
            npt.NDArray[np.floating]: Additive dispersion function [eta_x, eta_x', eta_y, eta_y'].
        '''
        kick_x, kick_y = self._tilted_kick(cood0.delta)
        eta_xp, eta_yp = -kick_x, -kick_y
        eta_x = -0.5 * kick_x * self._length
        eta_y = -0.5 * kick_y * self._length
        return np.array([eta_x, eta_xp, eta_y, eta_yp])

    def dispersion_array(self, cood0: Coordinate, ds: float = 0.1, endpoint: bool = False, method: str = 'symplectic4') \
        -> Tuple[npt.NDArray[np.floating], npt.NDArray[np.floating]]:
        '''
        Additive dispersion function along the steering magnet.

        Args:
            cood0 Coordinate: Initial coordinate.
            ds float: Maximum step size [m].
            endpoint bool: If True, include the endpoint.
            method str: Integration method. ('midpoint', 'rk4', 'symplectic{1,2,4}') (Not used in the Steering class.)

        Returns:
            npt.NDArray[np.floating]: Dispersion function array of shape (4, N).
            npt.NDArray[np.floating]: Longitudinal positions [m].
        '''
        s = self.s_array(ds, endpoint)
        kick_x, kick_y = self._tilted_kick(cood0.delta)
        if np.abs(self._length) > 0.:
            eta_xp = -kick_x * s / self._length
            eta_yp = -kick_y * s / self._length
            eta_x = -0.5 * kick_x * s**2 / self._length
            eta_y = -0.5 * kick_y * s**2 / self._length
        else:
            eta_xp = np.array([-kick_x])
            eta_yp = np.array([-kick_y])
            eta_x = np.array([0.])
            eta_y = np.array([0.])
        disp = np.array([eta_x, eta_xp, eta_y, eta_yp])
        return disp, s

    def transfer(self, cood0: Coordinate, evlp0: Envelope = None, disp0: Dispersion = None, ds: float = 0.1, method: str = 'symplectic4') \
        -> Tuple[Coordinate, Envelope, Dispersion]:
        '''
        Calculate the coordinate, envelope, and dispersion after the steering magnet.

        Args:
            cood0 Coordinate: Initial coordinate.
            evlp0 Envelope: Initial beam envelope (optional).
            disp0 Dispersion: Initial dispersion (optional).
            ds float: Maximum step size [m] for integration (not used in the Steering class).
            method str: Integration method. ('midpoint', 'rk4', 'symplectic{1,2,4}') (Not used in the Steering class.)

        Returns:
            Coordinate: Coordinate after the element.
            Envelope: Beam envelope after the element (if evlp0 is provided).
            Dispersion: Dispersion after the element (if disp0 is provided).
        '''
        kick_x, kick_y = self._tilted_kick(cood0.delta)
        dx, dy = 0.5 * self._length * kick_x, 0.5 * self._length * kick_y
        tmat = self.transfer_matrix(cood0, ds, method)
        cood = np.dot(tmat, cood0.vector) + np.array([dx, kick_x, dy, kick_y])
        cood1 = Coordinate(cood, cood0.s + self._length, cood0.z, cood0.delta)
        if evlp0 is not None:
            evlp1 = evlp0.copy()
            evlp1.transfer(tmat, self._length)
        else:
            evlp1 = None
        if disp0 is not None:
            disp = np.dot(tmat, disp0.vector) + self.dispersion(cood0, ds, method)
            disp1 = Dispersion(disp, disp0.s + self._length)
        else:
            disp1 = None
        return cood1, evlp1, disp1

    def transfer_array(self, cood0: Coordinate, evlp0: Envelope = None, disp0: Dispersion = None,
                       ds: float = 0.1, endpoint: bool = True, method: str = 'symplectic4') \
        -> Tuple[CoordinateArray, EnvelopeArray, DispersionArray]:
        '''
        Calculate the coordinate array along the steering magnet.

        Args:
            cood0 Coordinate: Initial coordinate.
            evlp0 Envelope: Initial beam envelope (optional).
            disp0 Dispersion: Initial dispersion (optional).
            ds float: Maximum step size [m].
            endpoint bool: If True, include the endpoint.
            method str: Integration method. ('midpoint', 'rk4', 'symplectic{1,2,4}') (Not used in the Steering class.)

        Returns:
            CoordinateArray: Coordinate array along the element.
            EnvelopeArray: Beam envelope array along the element (if evlp0 is provided).
            DispersionArray: Dispersion array along the element (if disp0 is provided).
        '''
        kick_x, kick_y = self._tilted_kick(cood0.delta)
        tmat, s = self.transfer_matrix_array(cood0, ds, endpoint, method)
        if np.abs(self._length) > 0.:
            x, y = 0.5 * s**2 * kick_x / self._length, 0.5 * s**2 * kick_y / self._length
            xp, yp = s * kick_x / self._length, s * kick_y / self._length
        else:
            x, y = np.array([0.]), np.array([0.])
            xp, yp = np.array([kick_x]), np.array([kick_y])
        cood = np.matmul(tmat, cood0.vector).T + np.array([x, xp, y, yp])
        cood1 = CoordinateArray(cood, s + cood0.s,
                                np.full_like(s, cood0.z), np.full_like(s, cood0.delta))
        if evlp0 is not None:
            evlp1 = EnvelopeArray.transport(evlp0, tmat, s)
        else:
            evlp1 = None
        if disp0 is not None:
            disp_add, _ = self.dispersion_array(cood0, ds, endpoint, method)
            disp = np.matmul(tmat, disp0.vector).T + disp_add
            disp1 = DispersionArray(disp, s + disp0.s)
        else:
            disp1 = None
        return cood1, evlp1, disp1

    def set_steering(self, kick_x: float | None = None, kick_y: float | None = None) -> None:
        '''
        Set the steering angles.

        Args:
            kick_x float: Horizontal deflection angle [rad].
            kick_y float: Vertical deflection angle [rad].
        '''
        if kick_x is not None:
            self._kick_x = kick_x
        if kick_y is not None:
            self._kick_y = kick_y
